### DISCUSSION TITLE
multiple lines in pdf extract as a sentence

### DISCUSSION THREAD

**Discussion by yonglee7015** [jsvine/pdfplumber#915] [Ask for help with specific PDFs]
![image](https://github.com/jsvine/pdfplumber/assets/59268746/581220ae-3bda-4c34-a806-06f475895b00)

pdfplumber extract texts result:
066-22C104978 - : CC SEQUIN TRIM ZIP CARDI
Semi-Worsted - 2/48NM_52% Cotton, 26% Rayon, 17% Nylon, 3% Spandex, 2%
Primary Material:
Cashmere: YAR-4522 - Cobalt_FOB
Vendor (Supplier): Cobalt_FOB
Season\Floor Set: Chicos Frontline 2022 Holiday, Holiday

How can I extract the text to: (any parameter to control it) 
**066-22C104978 - : CC SEQUIN TRIM ZIP CARDI
Primary Material:
Semi-Worsted - 2/48NM_52% Cotton, 26% Rayon, 17% Nylon, 3% Spandex, 2%
Cashmere: YAR-4522 - Cobalt_FOB**
Vendor (Supplier): Cobalt_FOB
Season\Floor Set: Chicos Frontline 2022 Holiday, Holiday

**Semi-Worsted - 2/48NM_52% Cotton, 26% Rayon, 17% Nylon, 3% Spandex, 2%
Cashmere: YAR-4522 - Cobalt_FOB**
 should be a sentence.

**Comment by cmdlineluser** [comment]
Are you using `.extract_text(layout=True)`?

It looks like the data may be a "table" - you could try using the `text` strategy of the table extraction methods.

    "vertical_strategy": "text", 
    "horizontal_strategy": "text",

If the data is public and you can share an example PDF - that would make it easier to help.

**Reply by yonglee7015** [reply]
Thank you for your help. I'm using .extract_text() without any parameters.  this is just a part of  the page. I am just wondering if there is any parameters in extract_text() to handle this without table extract. Such as space distance between two lines to judge if multiple lines is a sentence.

I know it always extract text from left to right, and from top to bottom.
I have so many different layout of pdf. this is just one type.

**Reply by cmdlineluser** [reply]
Ah, okay.

The first I'd probably try is `layout=True`

There's also `use_text_flow=True`, and the `[xy]_*` type args e.g. `x_tolerance`, `y_tolerance`

https://github.com/jsvine/pdfplumber#extracting-text

> All remaining **kwargs are passed to .extract_words(...) 

So you can look at `.extract_words()` for the list of arguments.

**Reply by yonglee7015** [reply]
If I set y_tolerance > 3 it will combine multiple lines by chars. Got headache. Still can't find a best way to handle multi lines

**Reply by cmdlineluser** [reply]
Is it possible you can provide a sample PDF and the code you're using?

**Reply by yonglee7015** [reply]
can you provide your email address for me? I can send my code and pdf file to you.

**Reply by yonglee7015** [reply]
I searched website, I found pdfminer.six can extract multiple lines as a sentence based on line margin. But there is a weak of pdfminer, it can not extract text 100% based on texts order in pdf. How to solve it?

**Reply by yonglee7015** [reply]
```
from pdfminer.layout import LAParams, LTTextBox
from pdfminer.pdfpage import PDFPage
from pdfminer.pdfinterp import PDFResourceManager
from pdfminer.pdfinterp import PDFPageInterpreter
from pdfminer.converter import PDFPageAggregator

fp = open('test.pdf', 'rb')
rsrcmgr = PDFResourceManager()
#line_margin
laparams = LAParams(line_margin=0.3)
device = PDFPageAggregator(rsrcmgr, laparams=laparams)
interpreter = PDFPageInterpreter(rsrcmgr, device)
pages = PDFPage.get_pages(fp)

for page in pages:
    texts =[]
    print(page.rotate)
    print('Processing next page...')
    interpreter.process_page(page)
    layout = device.get_result()
    for lobj in sorted(layout, key=lambda x: (x.y0,-x.x0),reverse = True):
        if isinstance(lobj, LTTextBox):
            x, y, text = lobj.bbox[0], lobj.bbox[3], lobj.get_text()
            texts.append(text.replace("\n"," "))
    print(texts)
```

**Comment by kuriakose1959** [comment]
[page4_mod.pdf](https://github.com/jsvine/pdfplumber/files/12250293/page4_mod.pdf)
 with pdfplumber.open(file_path\\page4_mod.pdf") as pdf:
        first_page = pdf.pages[0]
        text = first_page.extract_text(keep_blank_chars=True)
        lines = text.split("\n")
        print(lines)

I tried to convert the above statement into text by using pdfplumber. The second line in the Narration column is getting extended as a s a new line.
output:
02 Aug Transfer to xx1805 CommBank app 
brain furniture  325.00  $  $12,286.68 CR  
  05 Aug Transfer to xx1805 CommBank app  10.00  $  $12,276.68 CR  
  06 Aug Transfer to xx1805 CommBank app  40.00  $  $12,236.68 CR  
  07 Aug Transfer to xx1805 CommBank app  90.00  $  $12,146.68 CR  
  08 Aug Transfer to xx1805 CommBank app 
tickets  1,250.00  $  $10,896.68 CR  
  08 Aug Transfer to xx1805 CommBank app  20.00  $  $10,876.68 CR  
  09 Aug Transfer to xx1805 CommBank app  40.00  $  $10,836.68 CR  
  09 Aug Transfer from xx1805 CommBank app    $10.00  $10,846.68 CR  
   11 Aug Transfer from xx1805 CommBank app    $650.00  $11,496.68 CR  
Expected out put:
 02 Aug Transfer to xx1805 CommBank app brain furniture  325.00  $  $12,286.68 CR  
  05 Aug Transfer to xx1805 CommBank app  10.00  $  $12,276.68 CR  
  06 Aug Transfer to xx1805 CommBank app  40.00  $  $12,236.68 CR  
  07 Aug Transfer to xx1805 CommBank app  90.00  $  $12,146.68 CR  
  08 Aug Transfer to xx1805 CommBank app tickets  1,250.00  $  $10,896.68 CR  
  08 Aug Transfer to xx1805 CommBank app  20.00  $  $10,876.68 CR  
  09 Aug Transfer to xx1805 CommBank app  40.00  $  $10,836.68 CR  
  09 Aug Transfer from xx1805 CommBank app    $10.00  $10,846.68 CR  
  11 Aug Transfer from xx1805 CommBank app    $650.00  $11,496.68 CR

Please help:

**Reply by cmdlineluser** [reply]
You could probably open a new post for this if you're still stuck.

If you use `keep_blank_chars=True` for example, each of the "new lines" start with spaces.

One common approach is to rejoin the lines afterwards using `itertools.groupby`
```python
>>> with pdfplumber.open("Downloads/page4_mod.pdf") as pdf:
...    first_page = pdf.pages[0]
...    text = first_page.extract_text(keep_blank_chars=True)
...    groups = itertools.groupby(text.splitlines(), lambda line: line.startswith(' '))
...    lines = []
...    for newline, group in groups:
...        if newline:
...            lines.extend(group)
...        else:
...            lines[-1] += '\n'.join(group)
```
```
>>> lines
['  02 Aug Transfer to xx1805 CommBank app brain furniture  325.00  $  $12,286.68 CR  ',
 '  05 Aug Transfer to xx1805 CommBank app  10.00  $  $12,276.68 CR  ',
 '  06 Aug Transfer to xx1805 CommBank app  40.00  $  $12,236.68 CR  ',
 '  07 Aug Transfer to xx1805 CommBank app  90.00  $  $12,146.68 CR  ',
 '  08 Aug Transfer to xx1805 CommBank app tickets  1,250.00  $  $10,896.68 CR  ',
 '  08 Aug Transfer to xx1805 CommBank app  20.00  $  $10,876.68 CR  ',
 '  09 Aug Transfer to xx1805 CommBank app  40.00  $  $10,836.68 CR  ',
 '  09 Aug Transfer from xx1805 CommBank app    $10.00  $10,846.68 CR  ',
 '   11 Aug Transfer from xx1805 CommBank app    $650.00  $11,496.68 CR  ']
```

**Comment by kuriakose1959** [comment]
Thank you very much for your prompt reply.
Your valuable suggestion worked perfectly with files which bank statements which do not have security column as first column. This security column contains QR code printed vertically(sideways) in the first column. Here our suggested method of itertools logs out with error "out of range at lines[-1] += '\n'.join(group)"I tried to solve by converting pdf into word using pypdf library. In word, i deleted security(first column) column. Then, using word - export option converted file back to pdf.
Thus I could extract 99% data with extended lines(full narration)THANK YOU ONCE AGAIN

    On Saturday, August 5, 2023 at 06:53:37 PM GMT+5:30, cmdlineluser ***@***.***> wrote:  
 
 


You could probably open a new post for this if you're still stuck.

If you use keep_blank_chars=True for example, each of the "new lines" start with spaces.

One common approach is to rejoin the lines afterwards using itertools.groupby
>>> with pdfplumber.open("Downloads/page4_mod.pdf") as pdf:
...    first_page = pdf.pages[0]
...    text = first_page.extract_text(keep_blank_chars=True)
...    groups = itertools.groupby(text.splitlines(), lambda line: line.startswith(' '))
...    lines = []
...    for newline, group in groups:
...        if newline:
...            lines.extend(group)
...        else:
...            lines[-1] += '\n'.join(group)>>> lines
['  02 Aug Transfer to xx1805 CommBank app brain furniture  325.00  $  $12,286.68 CR  ',
 '  05 Aug Transfer to xx1805 CommBank app  10.00  $  $12,276.68 CR  ',
 '  06 Aug Transfer to xx1805 CommBank app  40.00  $  $12,236.68 CR  ',
 '  07 Aug Transfer to xx1805 CommBank app  90.00  $  $12,146.68 CR  ',
 '  08 Aug Transfer to xx1805 CommBank app tickets  1,250.00  $  $10,896.68 CR  ',
 '  08 Aug Transfer to xx1805 CommBank app  20.00  $  $10,876.68 CR  ',
 '  09 Aug Transfer to xx1805 CommBank app  40.00  $  $10,836.68 CR  ',
 '  09 Aug Transfer from xx1805 CommBank app    $10.00  $10,846.68 CR  ',
 '   11 Aug Transfer from xx1805 CommBank app    $650.00  $11,496.68 CR  ']
